tiledb/rest/capnp/query.cc

```diff
@@ -71,16 +78,68 @@ tiledb::sm::Status query_serialize(
       case tiledb::sm::SerializationType::CAPNP: {
         kj::Array<::capnp::word> protomessage = messageToFlatArray(message);
         kj::ArrayPtr<const char> message_chars = protomessage.asChars();
-        *serialized_string = new char[message_chars.size()];
+
+        *serialized_string = new char
+            [message_chars.size() + totalNumOfBytesInBuffers +
+             totalNumOfBytesInOffsets + totalNumOfBytesInVarBuffers];
+        // Copy capnp part of message
         memcpy(*serialized_string, message_chars.begin(), message_chars.size());
-        *serialized_string_length = message_chars.size();
+        // Compute total message length
+        *serialized_string_length =
+            message_chars.size() + totalNumOfBytesInBuffers +
+            totalNumOfBytesInOffsets + totalNumOfBytesInVarBuffers;
+
+        char* serialized_message = *serialized_string;
+        // char** serialized_string remains intact
+        serialized_message += static_cast<int>(message_chars.size());
+        // Iterate in attributes and copy buffers to message
+        auto bufferBuilder = query_builder.getBuffers();
+        for (uint64_t I = 0; I < bufferBuilder.size(); I++) {
+          auto attributeBufferHeader = bufferBuilder[I];
+          auto atributeName = attributeBufferHeader.getAttributeName().cStr();
+          auto offsetBufferSizeInBytes =
+              attributeBufferHeader.getOffsetBufferSizeInBytes();
+          if (offsetBufferSizeInBytes) {
+            // variable size attribute buffer
+            uint64_t* existingBufferOffset = nullptr;
+            uint64_t* existingBufferOffsetSize = nullptr;
+            void* existingBuffer = nullptr;
+            uint64_t* existingBufferSize = nullptr;
+            query->get_buffer(
+                atributeName,
+                &existingBufferOffset,
+                &existingBufferOffsetSize,
+                &existingBuffer,
+                &existingBufferSize);
+            memcpy(
+                serialized_message,
+                existingBufferOffset,
+                size_t(*existingBufferOffsetSize));
+            serialized_message += *existingBufferOffsetSize;
+            memcpy(
+                *serialized_string,
```

Comment:
Looks like you are overwriting the beginning of the buffer here -- should be using `serialized_message`.

Reply:
of course, thanks